Add Micronaut OpenAPI generator server guide

andriy-dmytruk opened the pull request.

andriy-dmytruk committed:
Add Micronaut OpenAPI generator server guide

andriy-dmytruk committed:
Fix build for the generated micronaut-openapi-generator-server by adding swagger-annotations dependancy

burtbeckwith commented on guides/micronaut-openapi-generator-server/metadata.json:
```diff
@@ -0,0 +1,18 @@
+  "tags": ["openapi", "server"],
+  "category": "Getting Started",
+  "languages": ["java"],
+  "buildTools": ["maven", "gradle"],
```
delete this line, no need to specify `buildTools` if you're using both Gradle and Maven

burtbeckwith commented on guides/micronaut-openapi-generator-server/metadata.json:
```diff
@@ -0,0 +1,18 @@
+  "category": "Getting Started",
+  "languages": ["java"],
+  "buildTools": ["maven", "gradle"],
+  "publicationDate": "2022-02-01",
```
be sure to update the date so it's close to the date when it's published on the live site

andriy-dmytruk replied:
Also I am not sure what category this should be put in. "Getting started" might not be the right one.

sdelamo replied:
Add a new category to the enum named "API"

burtbeckwith commented on guides/micronaut-openapi-generator-server/metadata.json:
```diff
@@ -0,0 +1,18 @@
+  "authors": ["Andriy Dmytruk"],
+  "tags": ["openapi", "server"],
+  "category": "Getting Started",
+  "languages": ["java"],
```
are you planning on adding Kotlin and Groovy source?

burtbeckwith replied:
nvm, only Java makes sense because the code is generated

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/resources/application.yml:
```diff
@@ -0,0 +1,18 @@
+context-path: "/"
+
+micronaut:
+    application:
```
indent with 2 spaces for consistency

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/resources/application.yml:
```diff
@@ -0,0 +1,18 @@
+    application:
+        name: openapi-micronaut
+    server:
+        port: 8080
```
no need to specify port 8080, it's the default

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/library/Application.java:
```diff
@@ -0,0 +1,9 @@
+
+import io.micronaut.runtime.Micronaut;
+
+public class Application {
```
delete this class. during the build, an application is created and your source code is copied into that project. if you need to replace an auto-generated class like this because you've changed something, include it in the guide so it overwrites the generated file, otherwise there's no need

burtbeckwith commented on buildSrc/src/main/java/io/micronaut/guides/feature/SwaggerAnnotations.java:
```diff
@@ -0,0 +1,12 @@
+
+import jakarta.inject.Singleton;
+
+import static io.micronaut.starter.build.dependencies.Scope.TEST;
```
delete this unused import

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+
+We will use the jar option. Download it from
+https://repo1.maven.org/maven2/org/openapitools/openapi-generator-cli/5.4.0/openapi-generator-cli-5.4.0.jar
+and store it in the folder that you want to use for this project.
```
use "directory" instead of "folder" (10 instances)

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+===============================
+All the options for installation are given at the https://openapi-generator.tech/docs/installation[OpenAPI Generator website].
+
+If you installed the generator with a packet manager or bash launcher script, simply run
```
packet -> package

burtbeckwith replied:
maybe show (or just add a doc link) how to install with a package manager or bash script?

andriy-dmytruk replied:
Added short description for Homebrew and bash script.
Not sure if that is needed, as the link above takes to all the options for installation (updated its name, too).

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+Openapi generator supports `.yaml` and `.json` file formats for the definition file.
+We will use YAML due its simplicity and human readability.
+
+In the folder to which you downloaded the OpenAPI generator CLI create a file named `library-definition.yaml` and open it with your
```
"to which" -> "where", also add a comma after CLI

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+    description: Search for books and add new ones
+----
+<1> The version that will be used for parsing.
+<2> The info object contains general information about the api to its reader.
```
api -> API

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+object will be defined later in `components/schemas` section of definition.
+<8> The `"400"` status code will be produced by Micronaut in case of bad request, like an incorrect type supplied or
+failed validation. Even though Micronaut handles it automatically and no implementation is needed on our side, we add it
+for a complete api specification.
```
api -> API

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+All the available types and their validations are described in
+link:https://swagger.io/docs/specification/data-models/data-types/["Data Models (Schemas)" OpenAPI guide].
+
+We will define another path with `POST` operation similarly to the way we created the first one:
```
"similarly to the way" -> "like"

andriy-dmytruk replied:
Removed the subordinate clause completely ("similarly to the way we created the first one" -> "")

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+<3> We write that `BookInfo` object is required to be in the request body.
+We reference the same `BookInfo` schema that we will define next.
+
+NOTE: If you want to read more about body definitions, you can read the
```
"If you want to" -> "To", "you can read" -> "see"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+
+----
+<1> We define the `BookInfo` schema inside then `components/schemas` section.
+From this schema a java class will be generated with the same `BookInfo` classname.
```
java -> Java, "classname" -> "class name"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+From this schema a java class will be generated with the same `BookInfo` classname.
+<2> We define all the properties of `BookInfo`, including required validation on them. An abbreviated form is used
+for some YAML lists and dictionaries to reduce the number of rows and simplify readability.
+<3> We reference another scheme to be used as a property.
```
scheme -> schema (multiple instances)

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+== Generating server API from template
+
+Now we will generate server API files from our definition.
+Open the terminal in the same folder as `libaray-definition.yaml` file and run the following command:
```
libaray -> library

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+We want all the controllers to be generated in the `example.micronaut.library.controller` package.
+<5> We want all the models (data models, like `BookInfo`) to be in `example.micronaut.library.model` package.
+<6> We want to use @build@ as build tool. The supported values are `gradle`, `maven` and `all`.
+If nothing is specified, both maven and gradle files are generated.
```
maven -> Maven, gradle -> Gradle

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+[NOTE]
+=========
+If you plan to change the definition file and regenerate files, consider setting the `-p generateControllerAsAbstract=true`
+parameter (We don't recommend doing it during this guide, though). In this case an abstract class will be generated
```
We -> we

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+=========
+If you plan to change the definition file and regenerate files, consider setting the `-p generateControllerAsAbstract=true`
+parameter (We don't recommend doing it during this guide, though). In this case an abstract class will be generated
+for the API, while all the logic needs to be implemented in a different class (that inherits the API abstract class).
```
inherits -> extends

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+If you plan to change the definition file and regenerate files, consider setting the `-p generateControllerAsAbstract=true`
+parameter (We don't recommend doing it during this guide, though). In this case an abstract class will be generated
+for the API, while all the logic needs to be implemented in a different class (that inherits the API abstract class).
+This way your changes won't be overwritten by generation, but API will be updated.
```
API -> the API

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+This way your changes won't be overwritten by generation, but API will be updated.
+=========
+
+After running OpenAPI generator CLI should output information about generated files.
```
"generator CLI" -> "generator, the CLI"

andriy-dmytruk replied:
"the OpenAPI generator CLI" is the thing that we run and that outputs information, so probably no comma. Not sure if there is a way to make the sentence more clear, though.

burtbeckwith replied:
How about "After running, the OpenAPI generator CLI will output information about generated files."

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+├── README.md
+└── ...
+----
+<1> The `docs/` folder contains automatically generated markdown files with documentation about your API.
```
markdown -> Markdown

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+└── ...
+----
+<1> The `docs/` folder contains automatically generated markdown files with documentation about your API.
+<2> The `Application.java` will start the Micronaut server with all the detected Controllers.
```
"The `Application.java` will start" -> "Starts", remove "all"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+----
+<1> The `docs/` folder contains automatically generated markdown files with documentation about your API.
+<2> The `Application.java` will start the Micronaut server with all the detected Controllers.
+<3> The `BooksController.java` is generated based on paths with `books` tag. It is generated in the package we specified
```
"The `BooksController.java`" -> "`BooksController`"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+    public BookInfo(String name, BookAvailability availability);
+----
+
+Along with that it has getters and setters for parameters and jackson serialization annotation.
```
jackson -> Jackson, annotation -> annotations

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+Along with that it has getters and setters for parameters and jackson serialization annotation.
+
+To add the required functionality to the server we will first create a service to manage books in our library.
+Create file `src/main/java/example/micronaut/library/BookService.java` with the following content:
```
"Create file `src/main/java/example/micronaut/library/BookService.java`" -> "Create `BookService`"

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/library/BookService.java:
```diff
@@ -0,0 +1,33 @@
+@Singleton // <1>
+public class BookService {
+    private static final List<BookInfo> BOOKS = new ArrayList<>(Arrays.asList(
+            new BookInfo("Alice's Advantures in Wonderland", BookAvailability.AVAILABLE)
```
Advantures -> Adventures

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/library/BookService.java:
```diff
@@ -0,0 +1,33 @@
+    ));
+
+    public void addBook(BookInfo bookInfo) {
+        // Add book to the list
```
don't comment the obvious

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+source:library/BookService[]
+callout:singleton[1]
+
+Now open the `BooksController.java`. The class `BooksController` should have 2 methods named the same
```
"Now open the `BooksController.java`. The class `BooksController` should have 2" -> "Now open `BooksController`; the class should have two"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+callout:singleton[1]
+
+Now open the `BooksController.java`. The class `BooksController` should have 2 methods named the same
+as the operations we created in the definition file. The method have Micronaut framework annotations describing the
```
method -> methods

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+as the operations we created in the definition file. The method have Micronaut framework annotations describing the
+required API. We will now write their bodies.
+
+First, import and inject a `BookService` instance as a class attribute:
```
"import and inject a `BookService` instance as a class attribute:" -> "inject `BookService`:"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+First, import and inject a `BookService` instance as a class attribute:
+source:library/controller/BooksController[tag=import]
+source:library/controller/BooksController[tag=inject,indent=0]
+<1> Micronaut will create an instance of the class and inject it here.
```
"Micronaut will create an instance of the class and inject it here" -> "Micronaut will inject the singleton instance of the bean here"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+
+And finally, we will implement the `addBook` method:
+source:library/controller/BooksController[tag=search,indent=0]
+<1> Call the desired function and return empty string, that to send an empty body in response.
```
"Call the desired function and return empty string, that to send an empty body in response." -> "Call the function and return an empty string to create a response with an empty body."

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+connection: keep-alive
+----
+
+== Testing Application
```
"Testing Application" -> "Testing the Application"

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/library/controller/BooksController.java:
```diff
@@ -0,0 +1,100 @@
+public class BooksController {
+
+    //tag::inject[]
+    @Inject // <1>
```
typically @Inject and a field like this is used for dependency injection in tests, but for application code it's best to use constructor injection with a final field to make it immutable:

```
    private final BookService bookService;

    public BooksController(BookService bookService) { // <1>
        this.bookService = bookService;
    }
```

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+<1> Return a Reactor publisher, that will return the result of search method.
+
+And finally, we will implement the `addBook` method:
+source:library/controller/BooksController[tag=search,indent=0]
```
"tag=search" -> "tag=addBook"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+
+As we have noticed previously, some files were generated as templates for tests.
+
+We will modify the `src/test/java/example/micronaut/library/controllerBooksControllerTest` to test our paths.
```
"the `src/test/java/example/micronaut/library/controllerBooksControllerTest`" -> "`BooksControllerTest`"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+As we have noticed previously, some files were generated as templates for tests.
+
+We will modify the `src/test/java/example/micronaut/library/controllerBooksControllerTest` to test our paths.
+For path `/search` 2 methods were generated: one for testing the `search` method itself, and another for testing
```
spell out numbers less than 10 (i.e. 2 throughout), use digits for numbers 10+

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+<3> Store the response body in a variable.
+<4> Add assertion: we expect to get 2 books containing `"Guide"` as response.
+
+We will also add a test to the `/add` path:
```
"We will also add a test to the `/add` path:" -> "We will also update the generated `addBookClientApiTest` for the `/add` path:"

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/test/java/example/micronaut/library/controller/BooksControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    //tag::addBook[]
+    @Test // <1>
+    void addBookClientApiTest() throws IOException {
```
remove `throws IOException`

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/test/java/example/micronaut/library/controller/BooksControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    //tag::search[]
+    @Test // <1>
+    void searchClientApiTest() throws IOException {
```
remove `throws IOException`

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/test/java/example/micronaut/library/controller/BooksControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+    BooksController controller;
+
+    /**
+     * This test is used to check that the api available to client through
```
api -> API (same for other test)

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/test/java/example/micronaut/library/controller/BooksControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+package example.micronaut.library.controller;
+
+import example.micronaut.library.model.BookAvailability;
```
I had to manually add an import for `BookAvailability`, it wasn't in the generated code - is this a bug?

andriy-dmytruk replied:
Yes, it will have to be changed in the generator. Thanks.

burtbeckwith commented on guides/micronaut-openapi-generator-server/java/src/test/java/example/micronaut/library/controller/BooksControllerTest.java:
```diff
@@ -0,0 +1,96 @@
+        List<BookInfo> body = (List<BookInfo>) response.body(); // <3>
+
+        // then
+        Assertions.assertEquals(HttpStatus.OK, response.status());
```
My preference for JUnit tests is to use static imports for `Assertions` methods - is this possible with the generated code?

sdelamo replied:
I have replaced the Controller test completely. It uses statics imports now

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+=== Add Security
+
+We could have defined our security requirements by adding a security scheme to the `library-definition.yaml` file.
+For example, we will add basic authentication:
```
"basic" -> "HTTP Basic"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+      type: http
+      scheme: basic
+----
+<1> Define a security scheme inside the `components/securitySchemes`. We want to use basic auth for authentication.
```
basic -> Basic

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+NOTE: You can read more about describing various authentication in the
+link:https://swagger.io/docs/specification/authentication/["Authentication and Authorization" OpenAPI guide].
+
+The generator will then annotate such endpoints with Micronaut`s
```
"Micronaut`s" -> "the"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+steps to create the `AuthenticationProvider` and appropriate tests.
+
+NOTE: You can also read link:https://micronaut-projects.github.io/micronaut-security/latest/guide/[Micronaut Security documentation]
+or link:https://micronaut.io/guides[Micronaut guides] about security to learn more about all
```
delete "all"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+
+=== Generate Micronaut Client
+
+You can generate Micronaut Client based on the same `library-definition.yaml` file.
```
"Micronaut Client" -> "a Micronaut client"

burtbeckwith commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -0,0 +1,497 @@
+    -o html-docs
+----
+
+common:helpWithMicronaut.adoc[]
```
Delete helpWithMicronaut.adoc, that's only for guides written by Object Computing

sdelamo committed:
use resource: to import yaml

andriy-dmytruk committed:
Update style with Burt's review suggestions

andriy-dmytruk committed:
Remove unused import

andriy-dmytruk committed:
Change package to example.micronaut

andriy-dmytruk committed:
Minor style changes

sdelamo committed:
delete logback.xml

sdelamo committed:
Refactor BookControllerTest

- don’t inject EmbeddedServer, it is not used in the test.
- don’t inject BookController, it is not used in the test.
- To fetch a list use `Argument.listOf`
- Don’t specify application-json as content type. It is the default.
- Don’t specify application-json as accept header value. It is the default.
- Use static imports for Assertions
- Use UriBuilder to build uris with query value parameters.
- Use HttpRequest.POST method which takes a string. Remove unncessary usage of UriTemplate.

sdelamo committed:
implement BookAvailabilityTest

sdelamo committed:
implement BookInfoTest

sdelamo committed:
curl command needs to be surrounded by quotes

Otherwise you may get

zsh: no matches found: localhost:8080/search?book-name=Guide

sdelamo committed:
link to installation section

sdelamo committed:
Instead of linking to the JAR link to the docs

That way the guide is simpler to maintain

sdelamo committed:
Merge branch 'main' into pr/728

sdelamo commented on guides/micronaut-openapi-generator-server/java/.openapi-generator-ignore:
```diff
@@ -0,0 +1,23 @@
+# OpenAPI Generator Ignore
```
Should we have this file as part of the complete solution? Or should we delete it?

sdelamo commented on guides/micronaut-openapi-generator-server/java/.openapi-generator/FILES:
```diff
@@ -0,0 +1,22 @@
+.gitignore
```
Should we have this file as part of the complete solution? Or should we delete it?

andriy-dmytruk replied:
This and the version file is not used by the generator in any way. It seems to be some kind of debug file or something to know what files were generated and by version in case user needs to know.
It can be skipped by user by setting the `-D generateMetadata=false` option.
We can delete it. But if someone will try running the generation themselves, they will have it.

sdelamo commented on buildSrc/src/main/resources/pom.xml:
```diff
@@ -145,5 +145,12 @@
 
+        <!-- Swagger Annotations -->
+        <dependency>
+            <groupId>io.swagger</groupId>
```
When you add the open api feature in Micronaut CLI or Micronaut Launch we add the following depenency: 

https://github.com/micronaut-projects/micronaut-starter/blob/3.3.x/starter-core/src/main/java/io/micronaut/starter/feature/other/OpenApi.java#L56

`io.swagger.core.v3:swagger-annotations`

Here you are using `io.swagger:swagger-annotations` instead. Which one should we use in the guide and in starter?

andriy-dmytruk replied:
It seems like `io.swagger.core.v3:swagger-annotations` is the newer version. Thanks.
I will have to change this in the generator.

sdelamo replied:
Can you update the guide to use `io.swagger.core.v3:swagger-annotations` in the meanwhile?

andriy-dmytruk replied:
We cannot change to `io.swagger.core.v3:swagger-annotations` now.
They have a lot of differences with the current version, and most imports do not work, e.g.:
`import io.swagger.annotations.ApiModel` -> `import io.swagger.v3.oas.annotations.media.Schema`.
Since it is in the model files, we cannot update the imports.

This will have to be changed in the generator. After that we can update the guide to not use `swagger-annotations` as a feature, but simply `openapi`.

sdelamo commented on buildSrc/src/main/java/io/micronaut/guides/feature/SwaggerAnnotations.java:
```diff
@@ -0,0 +1,10 @@
+@Singleton
+public class SwaggerAnnotations extends AbstractFeature {
+    public SwaggerAnnotations() {
+        super("swagger-annotations", "swagger-annotations");
```
I think it would be worth extracting a swagger feature from the OpenAPI feature. 

https://github.com/micronaut-projects/micronaut-starter/blob/3.3.x/starter-core/src/main/java/io/micronaut/starter/feature/other/OpenApi.java

sdelamo commented on guides/micronaut-openapi-generator-server/java/.openapi-generator/VERSION:
```diff
@@ -0,0 +1 @@
+5.4.0-SNAPSHOT
```
Should we have this file as part of the complete solution? Or should we delete it?

sdelamo commented on guides/micronaut-openapi-generator-server/java/README.md:
```diff
@@ -0,0 +1,23 @@
+# openapi-micronaut
```
Should we have this file as part of the complete solution? Or should we delete it?

sdelamo commented on guides/micronaut-openapi-generator-server/java/docs/controllers/BooksController.md:
```diff
@@ -0,0 +1,52 @@
+# BooksController
```
Should we have this file as part of the complete solution? Or should we delete it?

andriy-dmytruk replied:
This could not be generated at all by setting `-D modelDocs=false -D apiDocs=false`, but by default is generated.
I would keep so that the directory structure would be the same as user would get if they generate it.

**The same is true for all `/docs` files.**

sdelamo commented on guides/micronaut-openapi-generator-server/java/docs/models/BookAvailability.md:
```diff
@@ -0,0 +1,17 @@
+
+
+# BookAvailability
```
Should we have this file as part of the complete solution? Or should we delete it?

sdelamo commented on guides/micronaut-openapi-generator-server/java/docs/models/BookInfo.md:
```diff
@@ -0,0 +1,22 @@
+
+
+# BookInfo
```
Should we have this file as part of the complete solution? Or should we delete it?

sdelamo commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/BookService.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.stream.Collectors;
+
+@Singleton // <1>
+public class BookService {
```
I think we should save the `BookInfo` in a MySQL database with Flyway migrations and using Micronaut data JDBC. 

I think this will involve creating a `JdbcRepository` a `MappedEntity` and a flyway migration.

andriy-dmytruk replied:
Maybe we can add it as a next step (like I added some comments about security)?
I don't think it should be required to have MySQL database to complete the guide.

sdelamo commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/BookService.java:
```diff
@@ -0,0 +1,32 @@
+    ));
+
+    public void addBook(BookInfo bookInfo) {
+        BOOKS.add(bookInfo);
```
Should the addition of a book be synchronized?

sdelamo commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/controller/BooksController.java:
```diff
@@ -0,0 +1,102 @@
+    ) {
+        return Mono.fromCallable(() -> {
+            bookService.addBook(bookInfo);
+            return ""; // <1>
```
this feels a bit weird to return an empty string.

andriy-dmytruk replied:
I agree. But I think returning `null` will result in an error. And it is different if we change the method to return an `HTTPResponse` object - I suppose I will need to change generation to this. Is there another way?

sdelamo replied:
In a real world API I would probably return 201 created with show uri in the HTTP location header

sdelamo commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/model/BookAvailability.java:
```diff
@@ -0,0 +1,62 @@
+/*
```
I think we should attempt to generate this file as part of the guide generation.

sdelamo commented on guides/micronaut-openapi-generator-server/java/src/main/java/example/micronaut/model/BookInfo.java:
```diff
@@ -0,0 +1,195 @@
+/*
```
I think we should attempt to generate this file as part of the guide generation.

sdelamo commented on guides/micronaut-openapi-generator-server/metadata.json:
```diff
@@ -0,0 +1,17 @@
+  "apps": [
+    {
+      "name": "default",
+      "features": ["netty-server", "http-client", "security", "reactor", "swagger-annotations"]
```
I don't think you need to add `netty-server` or `http-client` features

sdelamo reviewed (changes requested):
Great work @andriy-dmytruk 

However, I think like us to expore if we could generate the guide files `BookAvailability` `BookInfo`  automatically given the Open API yaml definition.  

They have a core dependency or a gradle plugin which we may be able to leverage:

https://github.com/OpenAPITools/openapi-generator#12---artifacts-on-maven-central

Generating the files automatically will allow us to keep the guide almost up to date when new versions of OpenAPI Generator is released. 

Moreover, I have done sever commits @andriy-dmytruk Let me know if you have concerns with any of those commits.

sdelamo committed:
don't show paths in yaml

sdelamo committed:
add OpenAPI Generator Link and quote

sdelamo committed:
format windows command

andriy-dmytruk commented on guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc:
```diff
@@ -170,7 +170,14 @@ If you are using Windows command prompt, run:
+     -p controllerPackage=example.micronaut.controller  \
+     -p modelPackage=example.micronaut.model  \
+     -p build=@build@  \
+     -p test=junit
```
Unluckily this won't work on Windows. Windwos Power Shell uses `'` to command continuation to new line, while command prompt uses `^` (I don't know why they did it in such a way, but it seems like it cannot be formatted properly😕).

sdelamo replied:
This won't work in MacOS either. It is just to display it better in the guide. I understand our readers will be able to take this command and format it to their OS. 

Do you think we should revert this command and use a one liner?

andriy-dmytruk replied:
This command is exactly the same to the one in Linux, so we can either remove it (and let people reformat the Linux one to work on their system) or keep the one-liner, so that they can copy it and run without any issues.

sdelamo replied:
Let's keep one command only. We use this format `\` as a separator for readability do in other guides.

andriy-dmytruk committed:
Add OpenAPIGenerator to automatically generate guide sources with OpenAPI

andriy-dmytruk commented:
@sdelamo I started generating the model files automatically using dependancy to [OpenAPI Generator core](https://mvnrepository.com/artifact/org.openapitools/openapi-generator/5.4.0). Could you take a look at the current implementation and refactor if needed.?

I suppose it is best if `build.gradle` and `pom.xml` files are generated like before, therefore I run the OpenAPI generator before everything. This however has the disadvantage, that `application.yml` cannot be automatically generated as it would be overwritten. But I guess it is part of a larger question of also using OpenAPI generator in starter, and this should be fine for now.

andriy-dmytruk committed:
Minor changes: fix windows command, remove unused features, add BookService synchronization

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        configurator.setOutputDir(destination.getPath())
+
+        if (testFramework == SPOCK) {
+            configurator.addAdditionalProperty("test", "spock")
```
please extract constants for "test" and "spock"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        if (testFramework == SPOCK) {
+            configurator.addAdditionalProperty("test", "spock")
+        } else if (testFramework == JUNIT) {
+            configurator.addAdditionalProperty("test", "junit")
```
please extract constants for "test" and "junit"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+            configurator.addAdditionalProperty("test", "junit")
+        }
+        if (buildTool == GRADLE) {
+            configurator.addAdditionalProperty("build", "gradle")
```
please extract constants for "build" and "gradle"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        if (buildTool == GRADLE) {
+            configurator.addAdditionalProperty("build", "gradle")
+        } else if (buildTool == MAVEN) {
+            configurator.addAdditionalProperty("build", "maven")
```
please extract constants for "build" and "maven"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        } else if (buildTool == MAVEN) {
+            configurator.addAdditionalProperty("build", "maven")
+        }
+        configurator.addAdditionalProperty("controllerPackage", destinationPackage + ".controller")
```
please extract constants for "controllerPackage" and ".controller"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+            configurator.addAdditionalProperty("build", "maven")
+        }
+        configurator.addAdditionalProperty("controllerPackage", destinationPackage + ".controller")
+        configurator.addAdditionalProperty("apiPackage", destinationPackage + ".api")
```
please, extract constants for "apiPackage" and ".api"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        }
+        configurator.addAdditionalProperty("controllerPackage", destinationPackage + ".controller")
+        configurator.addAdditionalProperty("apiPackage", destinationPackage + ".api")
+        configurator.addAdditionalProperty("modelPackage", destinationPackage + ".model")
```
please extract constants for "modelPackage" and ".model"

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+import static io.micronaut.starter.options.BuildTool.GRADLE
+import static io.micronaut.starter.options.BuildTool.MAVEN
+
+@Singleton
```
No need to annotate this as `@Singleton`.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        } else if (testFramework == JUNIT) {
+            configurator.addAdditionalProperty("test", "junit")
+        }
+        if (buildTool == GRADLE) {
```
We currently don't generate Guides with the Gradle Kotlin DSL but in the future, we may. There is a convenience method in `BuildTool::isGradle` which should be used instead.

We should replace 

`if (buildTool == GRADLE) {`

with: 

`if (buildTool.isGradle) {`

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo commented on guides/micronaut-openapi-generator-server/metadata.json:
```diff
@@ -0,0 +1,21 @@
+      "features": ["security", "reactor", "swagger-annotations"],
+      "openAPIGeneratorConfig": {
+        "definitionFile": "src/main/resources/library-definition.yaml",
+        "generatorName": "java-micronaut-server"
```
I assume we are always going to use the generator of `java-micronaut-server` in guides. Thus, I think we can default to this.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

andriy-dmytruk replied:
Great

sdelamo commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -0,0 +1,70 @@
+        configurator.addAdditionalProperty("apiPackage", destinationPackage + ".api")
+        configurator.addAdditionalProperty("modelPackage", destinationPackage + ".model")
+
+        config.properties?.each {
```
in groovy you can do this more idiomatic by using: 

`.each { k, v ->`

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/b34e78642671134c75b7cab67d372e8ac40ccc21

sdelamo committed:
extract methods

sdelamo committed:
delete every generated file but sources

sdelamo commented:
> @sdelamo I started generating the model files automatically using dependancy to [OpenAPI Generator core](https://mvnrepository.com/artifact/org.openapitools/openapi-generator/5.4.0). Could you take a look at the current implementation and refactor if needed.?

It is looking good!

> I suppose it is best if `build.gradle` and `pom.xml` files are generated like before, therefore I run the OpenAPI generator before everything. This however has the disadvantage, that `application.yml` cannot be automatically generated as it would be overwritten. But I guess it is part of a larger question of also using OpenAPI generator in starter, and this should be fine for now.

I have done a commit to remove every file generated by the open api generator but the source files: 

https://github.com/micronaut-projects/micronaut-guides/pull/728/commits/a4825b40dfca6b792080a9ba623509756d9e039c

We have to revisit the Micronaut open api generator to see if we can tweak the application.yml which we are currenlty geneator. I want to understand the Jackson features we are adding by default.

sdelamo commented:
@andriy-dmytruk 

I've created a PR to 3.3.1 https://github.com/OpenAPITools/openapi-generator/pull/11569

andriy-dmytruk commented on buildSrc/src/main/groovy/io/micronaut/guides/OpenAPIGenerator.groovy:
```diff
@@ -67,4 +60,56 @@ class OpenAPIGenerator {
+        additionalProperties.put(PROPERTY_CONTROLLER_PACKAGE, destinationPackage + PACKAGE_CONTROLLER)
+        additionalProperties.put(PROPERTY_API_PACKAGE, destinationPackage + PACKAGE_API)
+        additionalProperties.put(PROPERTY_MODEL_PACKAGE, destinationPackage + PACKAGE_MODEL)
+        additionalProperties.put(PROPERTY_GENERATOR_NAME, GuideMetadata.OpenAPIGeneratorConfig.GENERATOR_JAVA_MICRONAUT_SERVER)
```
Should probably be `config.generatorName` instead of `GuideMetadata.OpenAPIGeneratorConfig.GENERATOR_JAVA_MICRONAUT_SERVER`.
There is also `java-micronaut-client` that can be used in the guides, so the property is needed.

sdelamo replied:
> Should probably be config.generatorName instead of GuideMetadata.OpenAPIGeneratorConfig.GENERATOR_JAVA_MICRONAUT_SERVER

You are right. Please, do the change.

sdelamo committed:
Change category use API

andriy-dmytruk committed:
Remove Windows command and adding generatorName as additional property (It is added before that separately).

andriy-dmytruk committed:
Changes to the openapi generator .adoc file. Description for curl commands.

andriy-dmytruk committed:
Add comments to the BookAvailabilityTest

andriy-dmytruk committed:
Refactor the model tests. Add description to BookInfoTest.

andriy-dmytruk committed:
Add Micronaut Data JDBC as next step. Minor refactor.

sdelamo committed:
Merge branch 'master' into openapi-generator
